### PR TITLE
Feature/libuv sockets

### PR DESCRIPTION
**Pull request by ademar**
Provides alternative `libuv` based async tcp sockets.

### CODE REVIEW

**Comment by haf** [comment]
Great, please continue to do awesome =). Also great that we have a test suite that properly checks its operation! Looking forward to seeing this move forward. [This is partially relevant](https://github.com/FStarLang/FStar/issues/269)

**Comment by panesofglass** [comment]
Exciting!

File: examples/Example/Program.fs [hunk 1]
```diff
@@ -149,6 +149,7 @@ let main argv =
       mimeTypesMap          = mimeTypes
       homeFolder            = None
       compressedFilesFolder = None
-      logger                = logger }
+      logger                = logger
+      tcpServerFactory      = new DefaultTcpServerFactory() }
```
Comment (haf) [inline examples/Example/Program.fs]: Does this mean it'll be easy to vary the TCP implementation? =)

File: src/Suave.LibUv/Tcp.fs [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+
+  member this.on_new_connection (server : IntPtr) (status: int) =
+
+    if status < 0 then
+          "New connection error: " +  (new string (uv_strerror(status))) |> Log.intern logger "Suave.Tcp.LibUvSocket.on_new_connection"
+    else
+
+      let client = createHandle <| uv_handle_size(uv_handle_type.UV_TCP)
+
+      let _ = uv_tcp_init(loop, client)
```
Comment (haf) [inline src/Suave.LibUv/Tcp.fs]: Check result value?

File: src/Suave.LibUv/Tcp.fs [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+
+    logger.Log LogLevel.Info <| fun _ ->
+        { path          = "Suave.Tcp.tcpIpServer"
+          trace         = TraceHeader.empty
+          message       = "listener started in " + (startData.ToString())
+          level         = LogLevel.Info
+          ``exception`` = None
+          tsUTCTicks    = Globals.utcNow().Ticks }
+
+    let _ = uv_run(loop, UV_RUN_DEFAULT)
```
Comment (haf) [inline src/Suave.LibUv/Tcp.fs]: Check result value?

File: src/Suave.LibUv/Tcp.fs [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+  let synchronizationContextCallback = createHandle <| uv_handle_size(uv_handle_type.UV_ASYNC)
+
+  let closeEvent = new ManualResetEvent(false)
+
+  let opsPool = createLibUvOpsPool maxConcurrentOps
+
+  member this.run () =
+    Thread.BeginThreadAffinity()
+    this.initLoop()
+    try
```
Comment (haf) [inline src/Suave.LibUv/Tcp.fs]: Check return values of the below?

File: src/Suave/Web.fs [hunk 5]
```diff
@@ -579,7 +579,7 @@ module internal ParsingAndControl =
             verbose (sprintf "Socket error while sending BAD_REQUEST, exiting: %A" err)
 
         | err ->
-          verbose (sprintf "Socket error while sending BAD_REQUEST, exiting: %A" err)
+          verbose (sprintf "Socket error while processing request, exiting: %A" err)
```
Comment (haf) [inline src/Suave/Web.fs]: +1

File: src/Suave/Web.fs [hunk 6]
```diff
@@ -654,10 +653,11 @@ let startWebServerAsync (config : SuaveConfig) (webpart : WebPart) =
     Path.Combine(ParsingAndControl.resolveDirectory config.compressedFilesFolder, "_temporary_compressed_files")
 
   // spawn tcp listeners/web workers
+
   let servers = 
     config.bindings
     |> List.map (SuaveConfig.toRuntime config homeFolder compressionFolder true
-                 >> ParsingAndControl.startWebWorkerAsync (config.bufferSize, config.maxOps) webpart)
+                 >> (fun runtime -> ParsingAndControl.startWebWorkerAsync (config.bufferSize, config.maxOps) webpart runtime (config.tcpServerFactory.create(config.logger, config.maxOps, config.bufferSize,runtime.matchedBinding))))
```
Comment (haf) [inline src/Suave/Web.fs]: Can we break this line up?

File: src/Suave.LibUv/Native.fs [hunk 7]
```diff
@@ -0,0 +1,194 @@
+﻿module Suave.LibUv.Native
```
Comment (haf) [inline src/Suave.LibUv/Native.fs]: Duplicated in LibUv.fs?

**Comment by ademar** [comment]
As it is now ALL tests pass on Linux and OS X, haven't tested Windows yet. 

Now need to figure out how incorporate testing the new `LibUvServerFactory` in our build process.
@haf Do you have any suggestions on how to achieve this ?

**Comment by haf** [comment]
The unit tests are values, so we could probably instantiate them with different tcp factories.

**Comment by ademar** [comment]
All tests are passing now cc @haf

**Comment by ademar** [comment]
@haf anything else you want to get done here ? Can we merge this?

**Comment by haf** [comment]
`creating proxy/GET /redirect returns 'redirect': Exception: System.DllNotFoundException: libuv.dll`

when running locally on my machine.

**Comment by ademar** [comment]
You need to install the dll.

**Comment by ademar** [comment]
`brew install libuv --universal`

**Comment by haf** [comment]
Let's check for it in the rake script... I'll see if I find a good way of checking.

**Comment by ademar** [comment]
Yeah, I did not touched the rake script;  the appveyor and travicCI scripts do install libuv.

**Comment by ademar** [comment]
Between I'm thinking we wont be shipping libuv with our nuget; the user needs to install it by themselves.

**Comment by haf** [comment]
I agree, I just want to make it easy to get started by running the rakescript... I have something almost done now.

**Comment by ademar** [comment]
pkg-config does not find libuv in my Mac, it does on Ubuntu ...

**Comment by haf** [comment]
I see, it finds it for me because I've configured the native dll path that it looks in .

**Comment by haf** [comment]
How do you install on linux?

**Comment by ademar** [comment]
I think I ended compiling it.

On Thu, Oct 22, 2015 at 11:41 AM, Henrik Feldt <notifications@github.com>
wrote:

> How do you install on linux?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SuaveIO/suave/pull/306#issuecomment-150266311>.
>



-- 
Ademar Gonzalez
+1-647-891-3606
http://ademar.name
https://github.com/ademar

**Comment by haf** [comment]
How about this?

**Comment by ademar** [comment]
That could work.

**Comment by ademar** [comment]
That check is not gonna work in Windows ..

**Comment by haf** [comment]
So there.

**Comment by haf** [comment]
Also, we'd need to merge it to master... Working on that.

**Comment by haf** [comment]
@ademar What do you think? Does the merge look OK? I think it's ready to merge back into master; I've bumped the version too, but I'll give you the pleasure of pressing merge :)

File: src/Suave.LibUv/Suave.LibUv.fsproj [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    </DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <ConsolePause>false</ConsolePause>
+    <GenerateTailCalls>true</GenerateTailCalls>
+    <PlatformTarget>
+    </PlatformTarget>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="mscorlib" />
+    <Reference Include="FSharp.Core" />
```
Comment (haf) [inline src/Suave.LibUv/Suave.LibUv.fsproj]: Should we use FSharp.Core nuget like everywhere else, here, too?

Comment (ademar) [inline src/Suave.LibUv/Suave.LibUv.fsproj]: Yes.

**Comment by haf** [comment]
Something about the merge has made the tests start failing again.

File: src/Suave/Sockets/SocketOp.fs [hunk 9]
```diff
@@ -15,6 +15,7 @@ type Error =
   /// For a HTTP socket user this means a response of '400 Bad Request', or for
   /// example WebSockets would abort the connection.
   | InputDataError of string
+  | ConnectionError of string
```
Comment (haf) [inline src/Suave/Sockets/SocketOp.fs]: @ademar Could you document this case?

**Comment by haf** [comment]
Ok, I'm pretty much done; waiting for appveyor to test on Windows, but I haven't figured out what caused the MissingMethodExceptions from the merge yet.

**Comment by ademar** [comment]
Thanks Henrik for the help here. 

Merging :)

**Comment by panesofglass** [comment]
This excites me beyond reason! Thank you!

**Comment by haf** [comment]
@panesofglass  =)

**Pull request merged**